### PR TITLE
Add vcd_org_vdc_template resource and data source

### PR DESCRIPTION
**Pull request by adambarreiro**
## Overview

This PR adds a new resource and data source `vcd_org_vdc_template`, which allows to manage and read VDC Templates, respectively.

The configuration of a VDC template is similar to a VDC, naturally. The main special peculiarity is the "Binding" concept, which is briefly explained at https://github.com/vmware/go-vcloud-director/pull/686 and put into practice here. In the code, we use a unique Binding ID for each field: External network, Gateway edge cluster and Services edge cluster. Like in UI, on updates they get simply regenerated.

Also, at some points there are special quirks, that are explained with comments, and if they affect the user, also in the documentation.

## Tests

- [x] New acceptance tests passed in VCD 10.5.1.1
- [x] New acceptance tests passed in VCD 10.6.0
- [x] Binary tests passed in VCD 10.5.1.1

## Related

The resource to instantiate VDC Templates is in PR #1280

### CODE REVIEW

File: vcd/resource_vcd_nsxv_firewall_rule.go [hunk 1]
```diff
@@ -880,6 +880,16 @@ func stringInSlice(str string, list []string) bool {
...
 }
 
+// stringInSlicePartially checks whether the input string contains any of the strings of the slice
+func stringInSlicePartially(str string, list []string) bool {
+	for _, v := range list {
+		if strings.Contains(str, v) {
+			return true
+		}
+	}
+	return false
```
Comment (dataclouder) [inline vcd/resource_vcd_nsxv_firewall_rule.go:890]:
```suggestion
	return slices.ContainsFunc(list, func(s string) bool {return strings.Contains(str, s) })
```
It should do the same

File: vcd/resource_vcd_org_vdc_template.go [hunk 2]
```diff
@@ -0,0 +1,772 @@
...
+			GatewayConfiguration:    gateway,
+			StorageProfile:          storageProfiles,
+			ThinProvision:           d.Get("enable_thin_provisioning").(bool),
+			FastProvisioningEnabled: d.Get("enable_fast_provisioning").(bool),
+		},
+	}
+
+	// Get the compute configuration. There are too many combinations to perform "smart validations" here,
+	// and given that VCD behaves well when input is incorrect, we skip them.
+	// Schema guarantees that there's exactly 1 item, so we can retrieve every field it directly.
```
Comment (dataclouder) [inline vcd/resource_vcd_org_vdc_template.go:460]:
```suggestion
	// Schema guarantees that there's exactly 1 item, so we can retrieve every field directly.
```

Comment (adambarreiro) [reply vcd/resource_vcd_org_vdc_template.go]: Fixed! I did a last minute change of that comment and ended with a leftover.

File: vcd/resource_vcd_nsxv_firewall_rule.go (hunk 1 above)
Comment (adambarreiro) [reply vcd/resource_vcd_nsxv_firewall_rule.go]: Thanks for the tip! Changed. It works (test passed).

File: website/docs/r/org_vdc_template.html.markdown [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+* `enable_thin_provisioning` - (Optional) If `true`, the VDCs instantiated from this template will have Thin provisioning enabled. Defaults to `false`
+* `edge_gateway` - (Optional) VDCs instantiated from this template will create a new Edge Gateway with the provided setup. Required if any `provider_vdc` block
+  has defined a `gateway_edge_cluster_id`. This **unique** block has the following properties:
+  * `name` - (Required) Name of the Edge Gateway
+  * `description` - (Optional) Description of the Edge Gateway
+  * `ip_allocation_count` - (Optional) Allocated IPs for the Edge Gateway. Defaults to 0
+  * `network_name` - (Required) Name of the network to create with the Edge Gateway
+  * `network_description` - (Optional) Description of the network to create with the Edge Gateway
+  * `network_gateway_cidr` - (Required) CIDR of the Edge Gateway for the created network
+  * `static_ip_pool` - (Required) **One block** with a single IP range (this is a constraint due to a bug in VCD) that have the following properties:
```
Comment (adambarreiro) [inline website/docs/r/org_vdc_template.html.markdown:124]: Heads up for reviewers too.
Confirmed that only one block can be set due to a bug. Should we mention like that in the docs?

File: website/docs/r/org_vdc_template.html.markdown [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+data "vcd_network_pool" "np1" {
+  name = "NSX-T Overlay 1"
+}
+
+resource "vcd_org_vdc_template" "tmpl" {
+  name               = "myTemplate"
+  tenant_name        = "myAwesomeTemplate"
+  description        = "Requires System privileges"
```
Comment (lvirbalas) [inline website/docs/r/org_vdc_template.html.markdown:44]: Nit: would read better overall if this line/param was moved one line above (ordering was like in argument reference docs below).

File: website/docs/r/org_vdc_template.html.markdown [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  * `memory_guaranteed` - (Required for `AllocationVApp`, `AllocationPool` or `Flex`) The percentage of the Memory guaranteed to be available to VMs running within the VDC instantiated from this template
+  * `elasticity` - (Required for `Flex`) True if compute capacity can grow or shrink based on demand
+  * `include_vm_memory_overhead` - (Required for `Flex`) True if the instantiated VDC includes memory overhead into its accounting for admission control
+* `storage_profile` - (Required) A block that defines a storage profile that the VDCs instantiated from this template will use. Must be **at least one**, which has the following properties:
+  * `name` - (Required) Name of Provider VDC storage profile to use for the VDCs instantiated from this template
+  * `default` - (Required) True if this is default storage profile for the VDCs instantiated from this template. Only **one** block should have this set to `true`
+  * `limit` - (Required) Storage limit for the VDCs instantiated from this template, in Megabytes. 0 means unlimited
```
Comment (lvirbalas) [inline website/docs/r/org_vdc_template.html.markdown:113]: Nit: in above params we state "in MB", but here "in Megabytes". What about making them consistent?

File: website/docs/r/org_vdc_template.html.markdown (hunk 3 above)
Comment (lvirbalas) [reply website/docs/r/org_vdc_template.html.markdown:124]: Yes, but I'd rephrase to something like this: "(this is a constraint as of VCD 10.x.x)"

File: website/docs/r/org_vdc_template.html.markdown (hunk 5 above)
Comment (adambarreiro) [reply website/docs/r/org_vdc_template.html.markdown]: Fixed!

File: website/docs/r/org_vdc_template.html.markdown (hunk 4 above)
Comment (adambarreiro) [reply website/docs/r/org_vdc_template.html.markdown:43]: Fixed!

File: website/docs/r/org_vdc_template.html.markdown (hunk 3 above)
Comment (adambarreiro) [reply website/docs/r/org_vdc_template.html.markdown]: Done

**Review by dataclouder — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -70,3 +71,5 @@ require (
 	google.golang.org/grpc v1.63.2 // indirect
 	google.golang.org/protobuf v1.34.0 // indirect
 )
+
+replace github.com/vmware/go-vcloud-director/v2 => github.com/adambarreiro/go-vcloud-director/v2 v2.17.0-alpha.1.0.20240613141201-9423b82d7ede
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply go.mod]: Done

**Review by Didainius — APPROVED** [review]
LGTM. Tests passed. I have configured some manual  clusters and noticed that not all of the templates that are configured in the test can be spawn due to some limitations, but I guess spinning up is the topic in https://github.com/vmware/terraform-provider-vcd/pull/1280 :)

File: website/docs/r/org_vdc_template.html.markdown [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  * `default` - (Required) True if this is default storage profile for the VDCs instantiated from this template. Only **one** block should have this set to `true`
+  * `limit` - (Required) Storage limit for the VDCs instantiated from this template, in MB. 0 means unlimited
+* `enable_fast_provisioning` - (Optional) If `true`, the VDCs instantiated from this template will have Fast provisioning enabled. Defaults to `false`
+* `enable_thin_provisioning` - (Optional) If `true`, the VDCs instantiated from this template will have Thin provisioning enabled. Defaults to `false`
+* `edge_gateway` - (Optional) VDCs instantiated from this template will create a new Edge Gateway with the provided setup. Required if any `provider_vdc` block
+  has defined a `gateway_edge_cluster_id`. This **unique** block has the following properties:
+  * `name` - (Required) Name of the Edge Gateway
+  * `description` - (Optional) Description of the Edge Gateway
+  * `ip_allocation_count` - (Optional) Allocated IPs for the Edge Gateway. Defaults to 0
+  * `network_name` - (Required) Name of the network to create with the Edge Gateway
```
[the remark this answers is not in the document]
Comment (Didainius) [reply website/docs/r/org_vdc_template.html.markdown:121]: Just a thought. Maybe not worth digging into field changes too much, but at least the comment could include `routed` word :)

File: website/docs/r/org_vdc_template.html.markdown [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  * `network_description` - (Optional) Description of the network to create with the Edge Gateway
+  * `network_gateway_cidr` - (Required) CIDR of the Edge Gateway for the created network
+  * `static_ip_pool` - (Required) **One block** with a single IP range (this is a constraint due to a bug in VCD 10.5+) that have the following properties:
+    * `start_address` - (Required) Start address of the IP range
+    * `end_address` - (Required) End address of the IP range
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply website/docs/r/org_vdc_template.html.markdown]: Mmm, it seems the docs rendered don't support three indentation levels. I have changed this one to remove the most inner bullets, turning it into a single paragraph. Not as visually nice, but at least it is more understandable.

File: website/docs/r/org_vdc_template.html.markdown (hunk 7 above)
[the remark this answers is not in the document]
Comment (adambarreiro) [reply website/docs/r/org_vdc_template.html.markdown]: Done, changed to routed in names, descriptions and documentation. Tests still pass, so I didn't break anything :)

File: vcd/resource_vcd_org_vdc_template_test.go [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+    limit   = {{.StorageProfileLimit}}
+  }
+  storage_profile {
+    name    = "{{.StorageProfile2}}"
+    default = false
+    limit   = 128
+  }
+
+  enable_fast_provisioning = true
+  enable_thin_provisioning        = true
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/resource_vcd_org_vdc_template_test.go]: Fixed!

**Review by lvirbalas — APPROVED** [review]
